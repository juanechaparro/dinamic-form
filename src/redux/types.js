const types = {
  UPDATE_FORM_FIELD: "UPDATE_FORM_FIELD",
  SET_FORM_FIELD_ERROR: "SET_FORM_FIELD_ERROR",
  SET_STEPS_DATA: "STEPS_DATA",
  SET_STEPS_PATHS: "STEPS_PATHS",
  UI_OPEN_MODAL: "uiOpenModal",
  UI_CLOSE_MODAL: "uiCloseModal",
  CLEAR_FORM: "CLEAR_FORM",
};
export default types;
