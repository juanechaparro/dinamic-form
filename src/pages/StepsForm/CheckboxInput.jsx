export const CheckboxInput = ({ options, fieldName, handleChange, value }) => {
  return (
    <>
      {options.map((option) => (
        <div key={option.value}>
          <input
            className="checkbox"
            type="checkbox"
            id={`${fieldName}_${option.value}`}
            name={fieldName}
            value={option.value}
            checked={value.includes(option.value)}
            onChange={handleChange}
          />
          <label
            className="checkbox-label"
            htmlFor={`${fieldName}_${option.value}`}
          >
            {option.name}
          </label>
        </div>
      ))}
    </>
  );
};
