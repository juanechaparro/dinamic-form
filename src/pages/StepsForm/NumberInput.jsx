export const NumberInput = ({ value, handleChange, options, numberError }) => {
  return (
    <div>
      <input
        type="number"
        value={value}
        onChange={handleChange}
        min={options[0]}
        max={options[1]}
        className={`form-input ${numberError ? "input-error" : ""}`}
      />
      {numberError && <span className="error-message">{numberError}</span>}
    </div>
  );
};
