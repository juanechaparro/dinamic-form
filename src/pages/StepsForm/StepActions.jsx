import { Link } from "react-router-dom";
import PropTypes from "prop-types";

export const StepActions = ({
  order,
  StepsPaths,
  stepsLength,
  nextStep,
  openModal,
}) => {
  return (
    <div className="button-container">
      {order > 1 ? (
        <Link className={"button button-prev"} to={StepsPaths[order - 1]}>
          <span className="button-icon">←</span> Atras
        </Link>
      ) : null}
      {order < stepsLength ? (
        <button className="button" onClick={nextStep}>
          Siguiente <span className="button-icon">→</span>
        </button>
      ) : null}
      <button className="button mt-10 mobile" onClick={openModal}>
        Abrir Resumen
      </button>
    </div>
  );
};

StepActions.propTypes = {
  order: PropTypes.number,
  StepsPaths: PropTypes.object,
  stepsLength: PropTypes.number,
  nextStep: PropTypes.func,
  openModal: PropTypes.func,
};
