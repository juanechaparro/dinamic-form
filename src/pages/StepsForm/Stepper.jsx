import "../../styles/Stepper.css";
import PropTypes from "prop-types";
export const Stepper = ({ currentStep, totalSteps }) => {
  return (
    <div className="stepper">
      {[...Array(totalSteps)].map((_, index) => {
        const stepNumber = index + 1;
        const isActive = stepNumber <= currentStep;
        const circleStyle = {
          backgroundColor: isActive ? "#8004fc" : "transparent",
          color: isActive ? "#fff" : "#8004fc",
        };

        return (
          <div className="step" key={stepNumber}>
            <div className="circle" style={circleStyle}>
              {isActive ? stepNumber : ""}
            </div>
            {stepNumber < totalSteps && (
              <div className={`line ${isActive ? "active" : ""}`} />
            )}
          </div>
        );
      })}
      <div className="step-label">{`${currentStep}/${totalSteps}`}</div>
    </div>
  );
};

Stepper.propTypes = {
  currentStep: PropTypes.number.isRequired,
  totalSteps: PropTypes.number.isRequired,
};
