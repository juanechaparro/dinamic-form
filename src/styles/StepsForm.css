.stepsForm_container{
    display: flex;
    justify-content: center;
    align-items: center;
    height: 90vh;
    margin:25px;
   
}
.steps-form_fields-container{
    display: flex;
    flex-direction: column;
    width: 450px;
}
.button {
    padding: 10px 20px;
    font-size: 16px;
    font-weight: bold;
    color: #fff;
    background-color: #8004fc;
    border: none;
    border-radius: 4px;
    cursor: pointer;
  }
  
  .button-prev {
    margin-right: 10px;
    background-color: #fff;
    color: #8004fc;
    border: 2px solid #8004fc;
  }
  
  .button-icon {
    margin-right: 5px;
  }

  @media (max-width: 768px) {
    .button {
        font-size: 14px;
      }
    
  }