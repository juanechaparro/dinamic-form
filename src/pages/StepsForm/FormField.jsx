import { CheckboxInput } from "./CheckboxInput";
import { NumberInput } from "./NumberInput";
import { useSelector } from "react-redux";

export const FormField = ({
  fieldName,
  value,
  type,
  options,
  validate,
  onFieldChange,
}) => {
  const fieldError = useSelector((state) => {
    const fieldState = state.form[`${fieldName}Error`]?.value;
    return fieldState ? fieldState : "";
  });
  const handleChange = (e) => {
    let newValue;
    if (type === "checkbox") {
      const checkboxValue = e.target.value;
      const isChecked = e.target.checked;
      if (isChecked) {
        newValue = [...value, checkboxValue];
      } else {
        newValue = value.filter((val) => val !== checkboxValue);
      }
    } else {
      newValue = e.target.value;
    }
    onFieldChange(fieldName, newValue);

    // if (validate) {
    //   const error = validate(newValue, type, options);
    //   onFieldChange(`${fieldName}Error`, error);
    // }
  };

  const renderInput = () => {
    switch (type) {
      case "checkbox":
        return (
          <CheckboxInput
            options={options}
            fieldName={fieldName}
            handleChange={handleChange}
            value={value}
          />
        );

      case "number":
        return (
          <NumberInput
            value={value}
            handleChange={handleChange}
            options={options}
            error={fieldError}
          />
        );
      default:
        return (
          <div>
            <input
              type={type}
              value={value}
              onChange={handleChange}
              className={`form-input ${fieldError ? "input-error" : ""}`}
            />
            {fieldError && <span className="error-message">{fieldError}</span>}
          </div>
        );
    }
  };

  return (
    <div>
      <label htmlFor={fieldName}>{fieldName}</label>
      {renderInput()}
    </div>
  );
};
