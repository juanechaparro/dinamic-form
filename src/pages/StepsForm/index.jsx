import { useDispatch, useSelector } from "react-redux";
import { FormField } from "./FormField";
import { Link, useNavigate } from "react-router-dom";
import { Summary } from "../../components/Summary";
import { uiOpenModal } from "../../redux/actions/ui";
import { SummaryModal } from "../../components/SummaryModal";
import { setFormFieldError, updateFormField } from "../../redux/actions/form";
import { validate } from "../../utils";
import "../../styles/StepsForm.css";
import Stepper from "./Stepper";
import Swal from "sweetalert2";
export const StepsForm = ({
  component,
  path,
  order,
  description,
  type,
  options,
  name,
}) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const fieldValue = useSelector((state) => {
    const fieldState = state.form[component]?.value;
    return fieldState ? fieldState : "";
  });
  const { StepsPaths, StepsPathsLoading } = useSelector((state) => state.form);
  const stepsLength = StepsPathsLoading ? 0 : Object.keys(StepsPaths).length;

  const onFieldChange = (fieldName, value) => {
    dispatch(updateFormField(fieldName, value));
  };
  const openModal = () => {
    dispatch(uiOpenModal());
  };
  //   dispatch(updateFormField(`${component}Error`, validation));
  const nextStep = () => {
    const validation = validate(fieldValue, type, options);
    if (validation) {
      dispatch(setFormFieldError(`${component}`, validation));
      Swal.fire({
        title: "Error!",
        text: validation,
        icon: "error",
        confirmButtonText: "Ok",
      });
      return;
    } else {
      dispatch(setFormFieldError(`${component}`, ""));
      navigate(StepsPaths[order + 1]);
    }
  };
  if (StepsPathsLoading) {
    return <div>Loading...</div>;
  } else
    return (
      <>
        {component === "SummaryStep" ? (
          <Summary lastStep={true} />
        ) : (
          <main className="stepsForm_container">
            <section className="steps-form_fields-container">
              <Stepper currentStep={order} totalSteps={stepsLength} />
              <SummaryModal />
              <FormField
                fieldName={component}
                description={description}
                type={type}
                options={options}
                name={name}
                value={fieldValue}
                validate={validate}
                onFieldChange={onFieldChange}
              />
              <div className="button-container">
                {order > 1 ? (
                  <Link
                    className={"button button-prev"}
                    to={StepsPaths[order - 1]}
                  >
                    <span className="button-icon">←</span> Atras
                  </Link>
                ) : null}
                {order < stepsLength ? (
                  <button className="button" onClick={nextStep}>
                    Siguiente <span className="button-icon">→</span>
                  </button>
                ) : null}
                <button className="button mt-10 mobile" onClick={openModal}>
                  Abrir Resumen
                </button>
              </div>
            </section>
            <section className="desktop">
              <Summary />
            </section>
          </main>
        )}
      </>
    );
};
