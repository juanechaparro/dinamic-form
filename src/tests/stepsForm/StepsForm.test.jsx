import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { StepsForm } from "../../pages/StepsForm";
import { store } from "../../redux/store";
import { MemoryRouter } from "react-router-dom";
describe("pruebas en StepsForm", () => {
  test("renderiza correctamente el componente", () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <StepsForm
            component="FullName"
            order={1}
            description="Nombre y apellidos del cliente"
            type="text"
            id={1}
            name="Nombre completo"
          />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
