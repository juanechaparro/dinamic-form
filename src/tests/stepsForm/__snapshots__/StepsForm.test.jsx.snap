// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en StepsForm renderiza correctamente el componente 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
